### ISSUE TITLE
eos_acls with replaced or overridden option doesn't change the configuration if ACL is undefined or empty on EOS devices

### ISSUE THREAD

**Issue by stopendy** [ansible-collections/arista.eos#353]
##### SUMMARY

`state: replaced` and `state: overridden` options of `eos_acls` module don't work if the ACL is undefined or ACEs are empty on the target EOS device.

If the ACL is undefined, Ansible saids `OK` not `Changed`, even if there are some configuration differences. If ACL is defined but empty, Ansible generates the following error.

```
fatal: [eos1]: FAILED! => {"changed": false, "module_stderr": "'aces'", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error"}
```

I tested the same way with `ios_acls`, and confirmed that Ansible changes IOS device's configuration when the ACL is undefined or empty.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

eos_acls module

##### ANSIBLE VERSION

```
/home/endy/.local/lib/python3.9/site-packages/paramiko/transport.py:236: CryptographyDeprecationWarning: Blowfish has been deprecated
  "class": algorithms.Blowfish,
ansible [core 2.12.5]
  config file = /home/endy/projects/ansible_eos/ansible.cfg
  configured module search path = ['/home/endy/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/endy/.local/lib/python3.9/site-packages/ansible
  ansible collection location = /b
  executable location = /home/endy/.local/bin/ansible
  python version = 3.9.7 (default, Sep 21 2021, 00:13:39) [GCC 8.5.0 20210514 (Red Hat 8.5.0-3)]
  jinja version = 3.1.2
  libyaml = True
```

- ansible-core 2.12.5

##### COLLECTION VERSION

```
# /home/endy/.local/lib/python3.9/site-packages/ansible_collections
Collection Version
---------- -------
arista.eos 3.1.0  

# /home/endy/.ansible/collections/ansible_collections
Collection Version
---------- -------
arista.eos 5.0.0  
```

- arista.eos 5.0.0
- ansible.netcommon 3.0.0
- ansible.utils 2.6.1

Note that this issue is also tested with the newest main branch of `arista.eos`

##### CONFIGURATION

```
```

- (empty)

##### OS / ENVIRONMENT

Target OS version: Arista vEOS-lab 4.26.5M

##### STEPS TO REPRODUCE

1. Create a playbook with `eos_acls` module and `state: overridden` option.

```yaml
- hosts: eos
  gather_facts: false

  tasks:
    - name: override acls
      arista.eos.eos_acls:
        config:
          - afi: ipv4
            acls:
              - name: test
                aces:
                  - sequence: 10
                    grant: permit
                    protocol: ip
                    source:
                      host: 192.168.0.1
                    destination:
                      any: true
        state: overridden
```

2. Ensure that the ACL named `test` hasn't been created on the EOS device.
3. Run the playbook and confirm Ansible generates `OK` result.
4. Log into EOS device, enter the configuration mode, and run the following commands to create an empty ACL.

```
ip access-list test
exit
```

5. Run the playbook again to confirm that the Ansible playbook fails.

##### EXPECTED RESULTS

The configuration should be overridden as defined in `config` option in the playbook.

Arista EOS device's configuration should have the following configuration after running the playbook.

```sh
# show running-config section ip access-list
ip access-list test
   10 permit ip host 192.168.0.1 any
```

##### ACTUAL RESULTS

The playbook generates `OK` or `Failed` result depending on the EOS device's ACL configuration. After running the playbook, the configuration hasn't been changed.

If `test` ACL is undefined on the EOS device, Ansible outputs the following.

```
# ansible-playbook eos.yml -vvvv

TASK [override acls] **********************
task path: /home/endy/projects/ansible_eos/test.yml:5
redirecting (type: connection) ansible.builtin.network_cli to ansible.netcommon.network_cli
Loading collection ansible.netcommon from /home/endy/.ansible/collections/ansible_collections/ansible/netcommon
<192.168.0.3> attempting to start connection
<192.168.0.3> using connection plugin ansible.netcommon.network_cli
Found ansible-connection at path /home/endy/.local/bin/ansible-connection
<192.168.0.3> local domain socket does not exist, starting it
<192.168.0.3> control socket path is /home/endy/.ansible/pc/956250bf8d
<192.168.0.3> redirecting (type: connection) ansible.builtin.network_cli to ansible.netcommon.network_cli
<192.168.0.3> Loading collection ansible.netcommon from /home/endy/.ansible/collections/ansible_collections/ansible/netcommon
<192.168.0.3> Loading collection arista.eos from /home/endy/.ansible/collections/ansible_collections/arista/eos
<192.168.0.3> local domain socket listeners started successfully
<192.168.0.3> loaded cliconf plugin ansible_collections.arista.eos.plugins.cliconf.eos from path /home/endy/.ansible/collections/ansible_collections/arista/eos/plugins/cliconf/eos.py for network_os arista.eos.eos
<192.168.0.3> ssh type is set to auto
<192.168.0.3> autodetecting ssh_type
<192.168.0.3> ssh type is now set to libssh
<192.168.0.3> 
<192.168.0.3> local domain socket path is /home/endy/.ansible/pc/956250bf8d
<192.168.0.3> Using network group action arista.eos.eos for arista.eos.eos_acls
<192.168.0.3> ANSIBLE_NETWORK_IMPORT_MODULES: enabled via connection option
<192.168.0.3> ANSIBLE_NETWORK_IMPORT_MODULES: found arista.eos.eos_acls  at /home/endy/.ansible/collections/ansible_collections/arista/eos/plugins/modules/eos_acls.py
<192.168.0.3> ANSIBLE_NETWORK_IMPORT_MODULES: running arista.eos.eos_acls
<192.168.0.3> ANSIBLE_NETWORK_IMPORT_MODULES: complete
ok: [eos1] => {
    "before": [],
    "changed": false,
    "commands": [],
    "invocation": {
        "module_args": {
            "config": [
                {
                    "acls": [
                        {
                            "aces": [
                                {
                                    "destination": {
                                        "address": null,
                                        "any": true,
                                        "host": null,
                                        "port_protocol": null,
                                        "subnet_address": null,
                                        "wildcard_bits": null
                                    },
                                    "fragment_rules": null,
                                    "fragments": null,
                                    "grant": "permit",
                                    "hop_limit": null,
                                    "line": null,
                                    "log": null,
                                    "protocol": "ip",
                                    "protocol_options": null,
                                    "remark": null,
                                    "sequence": 10,
                                    "source": {
                                        "address": null,
                                        "any": null,
                                        "host": "192.168.0.1",
                                        "port_protocol": null,
                                        "subnet_address": null,
                                        "wildcard_bits": null
                                    },
                                    "tracked": null,
                                    "ttl": null,
                                    "vlan": null
                                }
                            ],
                            "name": "test",
                            "standard": null
                        }
                    ],
                    "afi": "ipv4"
                }
            ],
            "running_config": null,
            "state": "overridden"
        }
    }
}
META: ran handlers
META: ran handlers
```

If `test` ACL is defined but empty on the EOS device, Ansible outputs the following.

```
# ansible-playbook -i hosts test.yml -vvvv

TASK [override acls] ********
task path: /home/endy/projects/ansible_eos/test.yml:5
redirecting (type: connection) ansible.builtin.network_cli to ansible.netcommon.network_cli
Loading collection ansible.netcommon from /home/endy/.ansible/collections/ansible_collections/ansible/netcommon
<192.168.0.3> attempting to start connection
<192.168.0.3> using connection plugin ansible.netcommon.network_cli
Found ansible-connection at path /home/endy/.local/bin/ansible-connection
<192.168.0.3> local domain socket does not exist, starting it
<192.168.0.3> control socket path is /home/endy/.ansible/pc/384d091222
<192.168.0.3> redirecting (type: connection) ansible.builtin.network_cli to ansible.netcommon.network_cli
<192.168.0.3> Loading collection ansible.netcommon from /home/endy/.ansible/collections/ansible_collections/ansible/netcommon
<192.168.0.3> Loading collection arista.eos from /home/endy/.ansible/collections/ansible_collections/arista/eos
<192.168.0.3> local domain socket listeners started successfully
<192.168.0.3> loaded cliconf plugin ansible_collections.arista.eos.plugins.cliconf.eos from path /home/endy/.ansible/collections/ansible_collections/arista/eos/plugins/cliconf/eos.py for network_os arista.eos.eos
<192.168.0.3> ssh type is set to auto
<192.168.0.3> autodetecting ssh_type
<192.168.0.3> ssh type is now set to libssh
<192.168.0.3> 
<192.168.0.3> local domain socket path is /home/endy/.ansible/pc/384d091222
<192.168.0.3> Using network group action arista.eos.eos for arista.eos.eos_acls
<192.168.0.3> ANSIBLE_NETWORK_IMPORT_MODULES: enabled via connection option
<192.168.0.3> ANSIBLE_NETWORK_IMPORT_MODULES: found arista.eos.eos_acls  at /home/endy/.ansible/collections/ansible_collections/arista/eos/plugins/modules/eos_acls.py
<192.168.0.3> ANSIBLE_NETWORK_IMPORT_MODULES: running arista.eos.eos_acls
<192.168.0.3> ANSIBLE_NETWORK_IMPORT_MODULES: complete
fatal: [eos1]: FAILED! => {
    "changed": false,
    "module_stderr": "'aces'",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error"
}
```

**Comment by stopendy** [comment]
If I should break this issue into 2 separate issues for undefined ACLs and empty ACLs, let me know and I'll create a new issue.

**Comment by stopendy** [comment]
Hi @GomathiselviS. 

Thank you very much for your prompt action! I could confirmed that this issue was fixed in all conditions:

- `state: overwritten` with undefined ACLs
- `state: overwritten` with empty ACLs
- `state: replaced` with undefined ACLs
- `state: replaced` with empty ACLs

I am so moved that you fixed this issue within a week with massive changes on source codes. This change allows me to use eos_acls even more reliably with `state: overwritten`!

So may I close this issue after I can confirm that #355 is merged?

**Comment by GomathiselviS** [comment]
Hello @stopendy ,

Glad that the PR fixed the issue. Once #355 is merged, this issue will close automatically, since I have linked the issue in the PR. I shall work on merging the PR.

**Comment by stopendy** [comment]
Hello @GomathiselviS,

Thank you for your advice! Sounds perfect. I look forward to seeing PR #355 available in general.
